### ISSUE TITLE
Should focusTrap only activate connected nodes?

### ISSUE THREAD

**Issue by bakura10** [focus-trap/focus-trap#724]
Hi,

First of all thanks for this library that made all our accessibility concerns much easier.

I am not sure if this is a bug or intended result, but I lost a few hours on this today. I am not completely convinced I fully understood the root cause of my issue, but what I did is that I have a piece of nodes that I create as a Fragment (so not yet in the DOM).

I actually manipulate them before inserting into the DOM, and among them, I have custom elements on which I change an attribute which, in turn, caused the "attributeChangedCallback" to run, which is the one on which I have my logic to trap the focus.

What seems to happen is that when I call activate method, the various listeners are added, but when I insert the fragment in the DOM, the nodes are different reference and the attributeChangedCallback is called again (and because the fragment nodes was never inserted, no "disconnectedCallback" is called to do any cleanup.

This therefore leaves a dangling reference somewhere which caused some hard to track issues.

Long story short, I have been able to solve the issue after hours of debugging by simply wrapping the activate code by this:

```
if (this.isConnected) {
   // trap focus
}
```

But actually, I wonder if this guard should not better be added in FocusTrap itself, just here: https://github.com/focus-trap/focus-trap/blob/master/index.js#L644

Is there actually any use case I did not think of where trapping focus on a node that is not connected makes sense?

Thanks :)

**Comment by stefcameron** [comment]
Hi, @bakura10 !

> First of all thanks for this library that made all our accessibility concerns much easier.
> 
> I am not sure if this is a bug or intended result, but I lost a few hours on this today. I am not completely convinced I fully understood the root cause of my issue, but what I did is that I have a piece of nodes that I create as a Fragment (so not yet in the DOM).

Bummer. That can be frustrating!

> I actually manipulate them before inserting into the DOM, and among them, I have custom elements on which I change an attribute which, in turn, caused the "attributeChangedCallback" to run, which is the one on which I have my logic to trap the focus.
> 
> What seems to happen is that when I call activate method, the various listeners are added, but when I insert the fragment in the DOM, the nodes are different reference and the attributeChangedCallback is called again (and because the fragment nodes was never inserted, no "disconnectedCallback" is called to do any cleanup.
> 
> This therefore leaves a dangling reference somewhere which caused some hard to track issues.

Right, memory leaks can be frustratingly hard to track down.

> Long story short, I have been able to solve the issue after hours of debugging by simply wrapping the activate code by this:
> 
> ```
> if (this.isConnected) {
>    // trap focus
> }
> ```
> 
> But actually, I wonder if this guard should not better be added in FocusTrap itself, just here: https://github.com/focus-trap/focus-trap/blob/master/index.js#L644
> 
> Is there actually any use case I did not think of where trapping focus on a node that is not connected makes sense?

I understand where you're coming from, but focus-trap has always made the assumption -- had the _expectation_ -- that the container elements it's given are "connected" to the document when the trap is activated.

Given the separation of concerns principal, I don't see it as focus-trap's responsibility to handle this, especially since, as the consumer of focus-trap, you have far more intimate knowledge of what your code is doing and what conditions are necessary to ensure the focus-trap's expectations are met. I also think that the expectation that nodes are in the document is a reasonable one that is made a lot of other libraries out there. If focus-trap were to handle "disconnected", dangling DOM fragments, I'd see that more as an exception to the rule that would be specifically supported rather than just implied.

So I'm leaning toward saying this should not be the trap's concern. It expects nodes are "connected" to the document when it gets activated.

**Comment by bakura10** [comment]
That makes a lot of sense, and it is pretty easy to fix on user-land anyway (once I understood it :p) so I am ok with my fix.

I will close, thanks a lot for the answer and details !

**Comment by stefcameron** [comment]
My pleasure! Glad you reached out; it will probably help out someone else who finds themselves in a similar situation.
